### PR TITLE
Java: Promote HashWithoutSalt query

### PR DESCRIPTION
**Pull request by joefarebrother**
Promotes Hash Without Salt query  from experimental. 

Covers cases in which a password is hashed using a `MessageDigest`, but is not concatenated with a salt, either directly or through multiple calls to methods like `MessageDigest.update`. 

Does not currently cover uses of other hashing methods, such as PBKDF2, though improvements to this coverage are planned.

### CODE REVIEW

File: java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+    conf.hasFlowPath(source, sink) and
+    source.getASuccessor*() = mid and
+    mid.getASuccessor*() = sink and
+    node = mid.getNode()
+  )
+}
+
+/** Holds if `node` is on a path of a MessageDigest object that is used exactly once. */
+private predicate messageDigestFlowPath(DataFlow::Node node) {
+  flowPathMidPoint(any(MessageDigestUsedOnceConfig c), node) and
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll:261]: Could the two configs be combined, always counting up to HashSalted but only sinking in the HashOne state? Can you give an example that requires two configs and these supporting predicates?

File: java/ql/src/Security/CWE/CWE-759/HashWithoutSalt.ql [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ *       external/cwe/cwe-759
+ */
+
+import java
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.security.HashWithoutSaltQuery
+
+from Expr pw, Expr hash
+where passwordHashWithoutSalt(pw, hash)
+select hash, "$@ is hashed without a salt.", pw, "This password"
```
Comment (smowton) [inline java/ql/src/Security/CWE/CWE-759/HashWithoutSalt.ql:18]: Note warning about missing security-severity

File: java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+private DataFlow::Node getArg(MethodAccess ma, int arg) {
+  arg >= 0 and
+  result.asExpr() = ma.getArgument(arg)
+  or
+  arg = -1 and
+  result.asExpr() = ma.getQualifier()
+  or
+  arg = -1 and
+  not ma.hasQualifier() and
+  result.(DataFlow::InstanceAccessNode).getInstanceAccess().isImplicitMethodQualifier(ma)
```
[the remark this answers is not in the document]
Comment (joefarebrother) [reply java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll:113]: `ArgumentNode` doesn't seem to be exported publically by the dataflow library.

File: java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+/** Gets the `arg`th parameter of `m`, where -1 is the qualifier, or a field read of the qualifier. */
+private DataFlow::Node getParam(Method m, int arg) {
+  result.(DataFlow::ParameterNode).isParameterOf(m, arg)
+  or
+  arg = -1 and
+  exists(FieldRead fr |
+    fr.getEnclosingCallable() = m and
+    (fr.hasQualifier() implies fr.getQualifier() instanceof ThisAccess) and
+    result.asExpr() = fr
+  )
```
[the remark this answers is not in the document]
Comment (joefarebrother) [reply java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll:125]: It is, but this term adds a field read step as well.

File: java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll (hunk 1 above)
Comment (joefarebrother) [reply java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll:261]: I have tried that; however the use-use flow that *doesn't* change the flow state is still there; e.g. if you have 
```
md.update(password);
md.update(salt);
md.digest():
```
then there is still flow at state `HashOne` to the sink, as well as the flow at state `HashSalted`.

File: java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll (hunk 3 above)
[the remark this answers is not in the document]
Comment (smowton) [reply java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll:113]: It just now occurs to me... are you sure the dataflow logic isn't accounting for wrapper functions for us?

For example, if we had `addToDigest(data, digest);` and `addToDigest(data, digest) { digest.update(data); }`, doesn't the post-update node of `digest` propagate in turn to the post-update node of the wrapper function? Can you show a case where simply letting the dataflow lib do its thing doesn't work?

File: java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll (hunk 4 above)
[the remark this answers is not in the document]
Comment (smowton) [reply java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll:125]: Can you give an example showing the difference?

File: java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll (hunk 1 above)
Comment (smowton) [reply java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll:261]: Interesting, have you asked Tom if there's a good workaround for this?

File: java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll (hunk 4 above)
[the remark this answers is not in the document]
Comment (joefarebrother) [reply java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll:125]: I want to count wrappers like:
```
class SHA256 {
  MessageDigest md;
  SHA256() { md = MessageDigest.getInstance("SHA256"); }
  void update(byte[] bs) { md.update(bs); }
  byte[] digest() { return md.digest(); }
}
```
which requires going through a field read of the `md` field.

File: java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll (hunk 3 above)
[the remark this answers is not in the document]
Comment (joefarebrother) [reply java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll:113]: I asked about this recently. 
[Is it OK to link to slack messages on public PRs? If not then I've linked it in the internal issue]
Essentially the dataflow library doesn't see flow into or out of the wrapper method.

[the remark this answers is not in the document]
Comment (smowton) [reply java/ql/lib/semmle/code/java/security/HashWithoutSaltQuery.qll:113]: OK then, for now replace the two `arg = -1` cases with `DataFlow::getInstanceArgument` which is public, and we should ask at a near future team meeting about making ArgumentNode or similar public.

**Comment by atorralba** [comment]
is there anything else blocking the merge, other than https://github.com/github/codeql/pull/8541#discussion_r834217496?

Needs a docs review maybe?

**Comment by joefarebrother** [comment]
There's a performance issue on jdk. The unusual dataflow logic and workarounds probably deserve a look at by somebody like @aschackmull too.

**Review by lucascosti — DISMISSED** [review]
The change-note looks good for docs 👍

**Comment by atorralba** [comment]
@aschackmull could you take a look at the dataflow logic in this query, and see if something can be improved?

**Comment by aschackmull** [comment]
> @aschackmull could you take a look at the dataflow logic in this query, and see if something can be improved?

There is certainly some very unusual flow going on here with a lot of ad-hoc moving parts, so I'm not too surprised that this has potential for poor performance. My gut feeling is that we'll need to aim for something less ad-hoc to get performance under control, but I'd need to dig a bit deeper to say more. Also, there are lots of parts in here that could have poor performance: have you identified the pain points?

**Comment by atorralba** [comment]
> have you identified the pain points?

@joefarebrother do you remember where the issue was in JDK regarding performance?

**Comment by joefarebrother** [comment]
Seems to be some memory pressure (lots of instances of "pausing/unpausing evaluation") within some of the dataflow stages. 

Currently trying removing various aspects of the query to see if any impact performance.

**Comment by aschackmull** [comment]
> Seems to be some memory pressure (lots of instances of "pausing/unpausing evaluation") within some of the dataflow stages.

That typically just means that the flow you're trying to calculate is huge - possibly due to a too ad-hoc flow config. The query involves 3 different configurations, I guess step 1 is to identify which of these are problematic.

**Comment by joefarebrother** [comment]
Looks like the main culprit is in adding all field reads as flow steps; restricting it to just field reads containing a MessageDigest improves the performance.

**Comment by joefarebrother** [comment]
Now it no longer times out and takes multiple hours, but DCA still indicates a 20% slowdown on JDK

**Comment by vlkl-sap** [comment]
Apologies for butting in; I came across this PR by chance. While storing unsalted password hashes is poor security, using typical message digests for password hashing is in itself fundamentally insecure. Message digests are designed for integrity and optimized for speedy evaluation, while dedicated password hashing functions are optimized for the opposite to prevent brute-force attacks.
Please consider scrapping this query and replacing it with a (much simpler) query testing for use of message digests for password hashing. That would annoy people who have to interoperate with legacy systems (Excel, etc.), but overall I think it's the right thing to do.
Thanks.

**Comment by joefarebrother** [comment]
Hi @vlkl-sap, thank you for your feedback and apologies for the delayed response.

We have decided to keep the current query in place, as it covers a known CWE and the work is already done. We plan to improve the coverage as well as the query help to recommend the use of more secure hashing algorithms. 

Seperately, we will also look at improving our coverage of hashing algorithm issues. We will have to investigate the best way of implemeting this to minimize the noisiness of alerts.

**Pull request closed without merging**